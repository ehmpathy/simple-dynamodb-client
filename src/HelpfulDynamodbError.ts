export enum SimpleDynamodbOperation {
  QUERY = 'query',
  PUT = 'put',
  DELETE = 'delete',
  WRITE_TRANSACTION = 'write transaction',
}
export class HelpfulDynamodbError extends Error {
  constructor({
    operation,
    error,
    input,
  }: {
    operation: SimpleDynamodbOperation;
    error: Error;
    input: any;
  }) {
    const message = `
Error found executing dynamodb ${operation}.

Found Error:
${error.name}: ${error.message}

Input:
${JSON.stringify(input, null, 2)}
    `.trim();
    super(message);
  }
}
