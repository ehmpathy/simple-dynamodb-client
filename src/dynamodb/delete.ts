import { DynamoDB } from 'aws-sdk';

import { getDocumentClient } from './client';

export interface RelevantDeleteInput {
  /**
   * The name of the table from which to delete the item.
   */
  TableName: DynamoDB.DocumentClient.TableName;
  /**
   * A map of attribute names to AttributeValue objects, representing the primary key of the item to delete. For the primary key, you must provide all of the attributes. For example, with a simple primary key, you only need to provide a value for the partition key. For a composite primary key, you must provide values for both the partition key and the sort key.
   */
  Key: DynamoDB.DocumentClient.Key;
  /**
   * A condition that must be satisfied in order for a conditional DeleteItem to succeed. An expression can contain any of the following:   Functions: attribute_exists | attribute_not_exists | attribute_type | contains | begins_with | size  These function names are case-sensitive.   Comparison operators: = | &lt;&gt; | &lt; | &gt; | &lt;= | &gt;= | BETWEEN | IN      Logical operators: AND | OR | NOT    For more information about condition expressions, see Condition Expressions in the Amazon DynamoDB Developer Guide.
   */
  ConditionExpression?: DynamoDB.DeleteItemInput['ConditionExpression'];
  /**
   * One or more values that can be substituted in an expression. Use the : (colon) character in an expression to dereference an attribute value. For example, suppose that you wanted to check whether the value of the ProductStatus attribute was one of the following:   Available | Backordered | Discontinued  You would first need to specify ExpressionAttributeValues as follows:  { ":avail":{"S":"Available"}, ":back":{"S":"Backordered"}, ":disc":{"S":"Discontinued"} }  You could then use these values in an expression, such as this:  ProductStatus IN (:avail, :back, :disc)  For more information on expression attribute values, see Condition Expressions in the Amazon DynamoDB Developer Guide.
   */
  ExpressionAttributeValues?: DynamoDB.DeleteItemInput['ExpressionAttributeValues'];
}

export const deleteItem = async ({ input }: { input: RelevantDeleteInput }) => {
  const dynamodbClient = getDocumentClient();
  return dynamodbClient
    .delete({
      // return consumed capacity by default
      ReturnConsumedCapacity: 'TOTAL',

      // where, limit, etc
      ...input,
    })
    .promise();
};
