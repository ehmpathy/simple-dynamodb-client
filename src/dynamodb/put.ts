import { DynamoDB } from 'aws-sdk';

export interface RelevantPutInput {
  /**
   * The name of the table to contain the item.
   */
  TableName: DynamoDB.DocumentClient.TableName;
  /**
   * A map of attribute name/value pairs, one for each attribute. Only the primary key attributes are required; you can optionally provide other attribute name-value pairs for the item. You must provide all of the attributes for the primary key. For example, with a simple primary key, you only need to provide a value for the partition key. For a composite primary key, you must provide both values for both the partition key and the sort key. If you specify any attributes that are part of an index key, then the data types for those attributes must match those of the schema in the table's attribute definition. Empty String and Binary attribute values are allowed. Attribute values of type String and Binary must have a length greater than zero if the attribute is used as a key attribute for a table or index. For more information about primary keys, see Primary Key in the Amazon DynamoDB Developer Guide. Each element in the Item map is an AttributeValue object.
   */
  Item: DynamoDB.DocumentClient.PutItemInputAttributeMap;
  /**
   * A condition that must be satisfied in order for a conditional PutItem operation to succeed. An expression can contain any of the following:   Functions: attribute_exists | attribute_not_exists | attribute_type | contains | begins_with | size  These function names are case-sensitive.   Comparison operators: = | &lt;&gt; | &lt; | &gt; | &lt;= | &gt;= | BETWEEN | IN      Logical operators: AND | OR | NOT    For more information on condition expressions, see Condition Expressions in the Amazon DynamoDB Developer Guide.
   */
  ConditionExpression?: DynamoDB.PutItemInput['ConditionExpression'];
  /**
   * One or more values that can be substituted in an expression. Use the : (colon) character in an expression to dereference an attribute value. For example, suppose that you wanted to check whether the value of the ProductStatus attribute was one of the following:   Available | Backordered | Discontinued  You would first need to specify ExpressionAttributeValues as follows:  { ":avail":{"S":"Available"}, ":back":{"S":"Backordered"}, ":disc":{"S":"Discontinued"} }  You could then use these values in an expression, such as this:  ProductStatus IN (:avail, :back, :disc)  For more information on expression attribute values, see Condition Expressions in the Amazon DynamoDB Developer Guide.
   */
  ExpressionAttributeValues?: DynamoDB.PutItemInput['ExpressionAttributeValues'];
}

export const putItem = async ({ input }: { input: RelevantPutInput }) => {
  const dynamodbClient = new DynamoDB.DocumentClient();
  return dynamodbClient.put(input).promise();
};
