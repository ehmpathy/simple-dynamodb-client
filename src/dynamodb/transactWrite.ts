import { DynamoDB } from 'aws-sdk';

export interface RelevantTransactWriteDeleteItemInput {
  /**
   * The primary key of the item to be deleted. Each element consists of an attribute name and a value for that attribute.
   */
  Key: DynamoDB.DocumentClient.Key;
  /**
   * Name of the table in which the item to be deleted resides.
   */
  TableName: DynamoDB.DocumentClient.TableName;
  /**
   * A condition that must be satisfied in order for a conditional delete to succeed.
   */
  ConditionExpression?: DynamoDB.DocumentClient.ConditionExpression;
  /**
   * One or more values that can be substituted in an expression.
   */
  ExpressionAttributeValues?: DynamoDB.DocumentClient.ExpressionAttributeValueMap;
}
export interface RelevantTransactWritePutItemInput {
  /**
   * A map of attribute name to attribute values, representing the primary key of the item to be written by PutItem. All of the table's primary key attributes must be specified, and their data types must match those of the table's key schema. If any attributes are present in the item that are part of an index key schema for the table, their types must match the index key schema.
   */
  Item: DynamoDB.DocumentClient.PutItemInputAttributeMap;
  /**
   * Name of the table in which to write the item.
   */
  TableName: DynamoDB.DocumentClient.TableName;
  /**
   * A condition that must be satisfied in order for a conditional update to succeed.
   */
  ConditionExpression?: DynamoDB.DocumentClient.ConditionExpression;
  /**
   * One or more values that can be substituted in an expression.
   */
  ExpressionAttributeValues?: DynamoDB.DocumentClient.ExpressionAttributeValueMap;
}

export type RelevantTransactWriteItemInput = { Put: RelevantTransactWritePutItemInput } | { Delete: RelevantTransactWriteDeleteItemInput };

export interface RelevantTransactWriteInput {
  /**
   * An ordered array of up to 25 TransactWriteItem objects, each of which contains a ConditionCheck, Put, Update, or Delete object. These can operate on items in different tables, but the tables must reside in the same AWS account and Region, and no two of them can operate on the same item.
   */
  TransactItems: RelevantTransactWriteItemInput[];

  /**
   * Determines whether consumed capacity information is returned;
   */
  ReturnConsumedCapacity?: DynamoDB.DocumentClient.ReturnConsumedCapacity;
}

export const transactWrite = async ({ input }: { input: RelevantTransactWriteInput }): Promise<DynamoDB.DocumentClient.TransactWriteItemsOutput> => {
  const dynamodbClient = new DynamoDB.DocumentClient();

  // define the request the request
  const transactionRequest = dynamodbClient.transactWrite({
    // return consumed capacity by default
    ReturnConsumedCapacity: 'TOTAL',

    // where, limit, etc
    ...input,
  });

  // add a event listener, to expose access to the "cancellation reasons", since the sdk does not expose them yet: https://github.com/aws/aws-sdk-js/issues/2464#issuecomment-503524701
  let cancellationReasons: any[];
  transactionRequest.on('extractError', (response) => {
    try {
      cancellationReasons = JSON.parse(response.httpResponse.body.toString()).CancellationReasons;
    } catch (err) {
      // suppress this just in case some types of errors aren't JSON parsable
    }
  });

  // now send the request - and if an error is caught, append the cancellation reasons to it (if any)
  return new Promise((resolve, reject) => {
    transactionRequest.send((err, response) => {
      if (err) {
        const errorMessage = cancellationReasons
          ? [err.message, '', 'Cancellation reasons:', JSON.stringify(cancellationReasons, null, 2)].join('\n')
          : err.message;
        const error = new Error(errorMessage);
        error.name = 'TransactionCanceledException'; // set the name to match the error name aws uses
        return reject(error);
      }
      return resolve(response);
    });
  });
};
