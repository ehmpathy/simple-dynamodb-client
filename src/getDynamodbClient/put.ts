import { DynamoDB } from 'aws-sdk';

export interface RelevantPutInput {
  Item: DynamoDB.DocumentClient.PutItemInputAttributeMap;
  ConditionExpression?: DynamoDB.DocumentClient.ConditionExpression;
}

export const putItem = async ({
  dynamodbClient,
  tableName,
  input,
}: {
  dynamodbClient: DynamoDB.DocumentClient;
  tableName: string;
  input: RelevantPutInput;
}) => {
  await dynamodbClient
    .put({
      // into table
      TableName: tableName,

      // item, conditions, etc
      ...input,
    })
    .promise();
};
