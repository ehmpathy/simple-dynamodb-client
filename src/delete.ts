import { DynamoDB } from 'aws-sdk';

import {
  HelpfulDynamodbError,
  SimpleDynamodbOperation,
} from './HelpfulDynamodbError';
import { dynamodb } from './dynamodb';
import { LogMethod } from './types';

export interface SimpleDynamodbDeleteConditions {
  ConditionExpression?: DynamoDB.DocumentClient.DeleteItemInput['ConditionExpression'];
  ExpressionAttributeValues?: DynamoDB.DocumentClient.DeleteItemInput['ExpressionAttributeValues'];
}

// note: we use the name "del" here because "delete" is a reserved keyword
export const del = async ({
  tableName,
  logDebug,
  key,
  deleteConditions,
}: {
  tableName: string;
  logDebug: LogMethod;
  key: DynamoDB.DocumentClient.Key;
  deleteConditions?: SimpleDynamodbDeleteConditions;
}) => {
  try {
    logDebug(`${tableName}.delete.input`, {
      tableName,
      key,
      conditions: deleteConditions,
    });
    const response = await dynamodb.delete({
      input: {
        TableName: tableName,
        Key: key, // primary key of item to delete
        ConditionExpression: deleteConditions?.ConditionExpression,
        ExpressionAttributeValues: deleteConditions?.ExpressionAttributeValues,
      },
    });
    logDebug(`${tableName}.delete.output`, {
      success: true,
      tableName,
      key,
      conditions: deleteConditions,
      stats: {
        consumedCapacity: response.ConsumedCapacity,
      },
    });
  } catch (error) {
    if (!(error instanceof Error)) throw error;
    throw new HelpfulDynamodbError({
      operation: SimpleDynamodbOperation.DELETE,
      error,
      input: { tableName, key, deleteConditions },
    });
  }
};
