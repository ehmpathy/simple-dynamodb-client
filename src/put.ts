import { DynamoDB } from 'aws-sdk';

import {
  HelpfulDynamodbError,
  SimpleDynamodbOperation,
} from './HelpfulDynamodbError';
import { dynamodb } from './dynamodb';
import { LogMethod } from './types';

export interface SimpleDynamodbPutConditions {
  ConditionExpression?: DynamoDB.DocumentClient.PutItemInput['ConditionExpression'];
  ExpressionAttributeValues?: DynamoDB.DocumentClient.PutItemInput['ExpressionAttributeValues'];
}

export const put = async ({
  tableName,
  logDebug,
  item,
  putConditions,
}: {
  tableName: string;
  logDebug: LogMethod;
  item: object;
  putConditions?: SimpleDynamodbPutConditions;
}) => {
  try {
    logDebug(`${tableName}.put.input`, {
      tableName,
      item,
      conditions: putConditions,
    });
    const response = await dynamodb.put({
      input: {
        TableName: tableName,
        Item: item, // the item itself
        ConditionExpression: putConditions?.ConditionExpression,
        ExpressionAttributeValues: putConditions?.ExpressionAttributeValues,
      },
    });
    logDebug(`${tableName}.put.output`, {
      success: true,
      tableName,
      item,
      conditions: putConditions,
      stats: {
        consumedCapacity: response.ConsumedCapacity,
      },
    });
  } catch (error) {
    if (!(error instanceof Error)) throw error;
    throw new HelpfulDynamodbError({
      operation: SimpleDynamodbOperation.PUT,
      error,
      input: { tableName, item, putConditions },
    }); // make error more helpful when thrown
  }
};
