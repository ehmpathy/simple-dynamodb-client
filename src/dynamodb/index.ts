import { deleteItem } from './delete';
import { putItem } from './put';
import { getItem } from './get';
import { query } from './query';
import { transactWrite } from './transactWrite';

export const dynamodb = {
  delete: deleteItem,
  put: putItem,
  get: getItem,
  query,
  transactWrite,
};
