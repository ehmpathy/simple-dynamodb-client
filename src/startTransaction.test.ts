import { HelpfulDynamodbError } from './HelpfulDynamodbError';
import { del } from './delete';
import { deleteItem } from './dynamodb/delete';
import { putItem } from './dynamodb/put';
import { transactWrite } from './dynamodb/transactWrite';
import { put } from './put';
import { startTransaction } from './startTransaction';

const logDebug = () => jest.fn();

jest.mock('./dynamodb/put');
const putItemMock = putItem as jest.Mock;
putItemMock.mockReturnValue({ ConsumedCapacity: '__CONSUMED_CAPACITY__' });
jest.mock('./dynamodb/delete');
const deleteItemMock = deleteItem as jest.Mock;
deleteItemMock.mockReturnValue({ ConsumedCapacity: '__CONSUMED_CAPACITY__' });
jest.mock('./dynamodb/transactWrite');
const transactWriteMock = transactWrite as jest.Mock;
transactWriteMock.mockReturnValue({
  ConsumedCapacity: '__CONSUMED_CAPACITY__',
});

describe('beginWriteTransaction', () => {
  it('should be possible to queue a put', async () => {
    const transaction = startTransaction();
    transaction.queue.put({
      tableName: 'spaceships',
      item: { fuel: 9000 },
      logDebug,
    });
  });
  it('should be possible to interchange queue.put and normal put', async () => {
    const input = { tableName: 'spaceships', item: { fuel: 9000 }, logDebug };

    // input works for normal request
    await put(input);

    // input works for transaction too
    const transaction = startTransaction();
    transaction.queue.put(input);
  });

  it('should be possible to queue a delete', async () => {
    const transaction = startTransaction();
    transaction.queue.delete({
      tableName: 'spaceships',
      key: { p: 'uuid' },
      logDebug,
    });
  });
  it('should be possible to interchange queue.put and normal put', async () => {
    const input = { tableName: 'spaceships', key: { p: 'uuid' }, logDebug };

    // input works for normal request
    await del(input);

    // input works for transaction too
    const transaction = startTransaction();
    transaction.queue.delete(input);
  });

  it('should accurately execute a transaction', async () => {
    const transaction = startTransaction();
    transaction.queue.put({
      tableName: 'spaceships',
      item: { id: 821, fuel: 9000 },
      logDebug,
    });
    transaction.queue.put({
      tableName: 'spaceport',
      item: { spaceships: [{ id: 821 }] },
      logDebug,
    });
    transaction.queue.delete({
      tableName: 'cargo-to-spaceship',
      key: { p: 'SOIL', s: 821 },
      logDebug,
    });
    transaction.queue.put({
      tableName: 'cargo-to-spaceship',
      item: { p: 'SOIL', s: 721, quantity: 7 },
      logDebug,
    });
    await transaction.execute({ logDebug });
    expect(transactWriteMock).toHaveBeenCalledTimes(1);
    expect(transactWriteMock).toHaveBeenCalledWith({
      input: {
        TransactItems: [
          {
            Put: {
              TableName: 'spaceships',
              Item: { id: 821, fuel: 9000 },
              ConditionExpression: undefined,
              ExpressionAttributeValues: undefined,
            },
          },
          {
            Put: {
              TableName: 'spaceport',
              Item: { spaceships: [{ id: 821 }] },
              ConditionExpression: undefined,
              ExpressionAttributeValues: undefined,
            },
          },
          {
            Delete: {
              TableName: 'cargo-to-spaceship',
              Key: {
                p: 'SOIL',
                s: 821,
              },
              ConditionExpression: undefined,
              ExpressionAttributeValues: undefined,
            },
          },
          {
            Put: {
              TableName: 'cargo-to-spaceship',
              Item: {
                p: 'SOIL',
                s: 721,
                quantity: 7,
              },
              ConditionExpression: undefined,
              ExpressionAttributeValues: undefined,
            },
          },
        ],
      },
    });
  });
  it('should throw a helpful error if the transaction failed', async () => {
    transactWriteMock.mockRejectedValueOnce(
      new Error(
        'Transaction cancelled, please refer cancellation reasons for specific reasons [None, ConditionalCheckFailed]',
      ), // typical example message
    );
    const transaction = startTransaction();
    transaction.queue.put({
      tableName: 'spaceships',
      item: { id: 821, fuel: 9000 },
      logDebug,
    });
    transaction.queue.delete({
      tableName: 'cargo-to-spaceship',
      key: { p: 'SOIL', s: 821 },
      logDebug,
    });
    try {
      await transaction.execute({ logDebug });
      throw new Error('should not reach here');
    } catch (error) {
      if (!(error instanceof Error)) throw error;
      expect(error).toBeInstanceOf(HelpfulDynamodbError);
      expect(error.message).toContain(
        'Error: Transaction cancelled, please refer cancellation reasons for specific reasons [None, ConditionalCheckFailed]',
      );
      expect(error.message).toContain('"Put": {');
      expect(error.message).toContain('"Delete": {');
      expect(error.message).toMatchSnapshot(); // save an example for docs
    }
  });
  it('should be possible to get a static, synced timestamp to use for writes in the transaction', async () => {
    const transaction = startTransaction();
    transaction.queue.put({
      tableName: 'nature-preserve',
      item: {
        name: 'the great preserve',
        effectiveAt: transaction.startTimestamp,
      },
      logDebug,
    });
    transaction.queue.put({
      tableName: 'species',
      item: {
        kingdom: 'animal',
        class: 'mammals',
        family: 'foxes',
        species: 'silver fox',
      },
      logDebug,
    });
    transaction.queue.put({
      tableName: 'funding',
      item: {
        grant: 1000000,
        remaining: 50000,
        effectiveAt: transaction.startTimestamp,
      },
      logDebug,
    });
    await transaction.execute({ logDebug });
  });
});
