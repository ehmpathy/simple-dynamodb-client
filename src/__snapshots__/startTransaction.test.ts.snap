// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`beginWriteTransaction should throw a helpful error if the transaction failed 1`] = `
"An error was found attempting to execute a dynamodb transaction: Transaction cancelled, please refer cancellation reasons for specific reasons [None, ConditionalCheckFailed]

Here is the transaction that had the error:
[
  {
    \\"Put\\": {
      \\"TableName\\": \\"spaceships\\",
      \\"Item\\": {
        \\"id\\": 821,
        \\"fuel\\": 9000
      }
    }
  },
  {
    \\"Delete\\": {
      \\"TableName\\": \\"cargo-to-spaceship\\",
      \\"Key\\": {
        \\"p\\": \\"SOIL\\",
        \\"s\\": 821
      }
    }
  }
]"
`;
